hostsfile/src/lib.rs

```diff
@@ -179,24 +174,19 @@ impl HostsBuilder {
                 lines.len()
             },
             _ => {
-                return Err(Box::new(Error(format!(
-                    "start or end marker missing in {:?}",
-                    &hosts_path
-                ))));
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("start or end marker missing in {:?}", &hosts_path),
+                ));
             },
         };
 
-        // The tempfile should be in the same filesystem as the hosts file.
-        let hosts_dir = hosts_path
-            .parent()
-            .expect("hosts file must be an absolute file path");
-        let temp_dir = tempfile::Builder::new().tempdir_in(hosts_dir)?;
-        let temp_path = temp_dir.path().join("hosts");
-
-        // Copy the existing hosts file to preserve permissions.
-        fs::copy(&hosts_path, &temp_path)?;
-
-        let mut file = File::create(&temp_path)?;
+        let mut file = OpenOptions::new()
```

Comment:
Do you want to write all the contents into a `String` first and `write!` that to the file? It should be atomic in most operating systems according to:
https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.append

Hosts file shouldn't be edit frequently, and race conditions are not realistic, but since this is a specific crate for editing the file, I think it makes sense to give a bit of consideration to concurrent writes. Another option would be to use a lock file.

Reply:
That makes sense, thanks for looking!